File: EventOrganizer.java
// EventOrganizer.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.*;

class Task {
    String id;
    String description;
    String category;
    boolean completed;

    Task() {}
    Task(String description, String category) {
        this.id = UUID.randomUUID().toString().substring(0,8);
        this.description = description;
        this.category = category;
        this.completed = false;
    }
}

class Event {
    String id;
    String name;
    String date;
    String description;
    List<Task> tasks = new ArrayList<>();

    Event() {}
    Event(String name, String date, String description) {
        this.id = UUID.randomUUID().toString().substring(0,8);
        this.name = name;
        this.date = date;
        this.description = description;
    }
}

class Organizer {
    private List<Event> events = new ArrayList<>();
    private final String dataFile = "events.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public Organizer() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Event[] arr = gson.fromJson(json, Event[].class);
                events = Arrays.asList(arr);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(events).getBytes());
        } catch (Exception e) {}
    }

    private Event getEvent(String id) {
        for (Event e : events) if (e.id.equals(id)) return e;
        return null;
    }

    public void create(String name, String date, String desc) {
        Event e = new Event(name, date, desc);
        events.add(e);
        save();
        System.out.printf("✅ Event created: %s (ID: %s)%n", e.name, e.id);
    }

    public void addTask(String eventId, String desc, String category) {
        Event e = getEvent(eventId);
        if (e == null) {
            System.out.printf("Event %s not found.%n", eventId);
            return;
        }
        Task t = new Task(desc, category);
        e.tasks.add(t);
        save();
        System.out.printf("✅ Task added: %s (ID: %s)%n", t.description, t.id);
    }

    public void list() {
        if (events.isEmpty()) {
            System.out.println("No events.");
            return;
        }
        System.out.println("\n📋 Events:");
        for (Event e : events) {
            int total = e.tasks.size();
            int done = 0;
            for (Task t : e.tasks) if (t.completed) done++;
            int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
            System.out.printf("  %s (ID: %s) – %d/%d tasks done (%d%%)%n", e.name, e.id, done, total, pct);
        }
    }

    public void show(String eventId) {
        Event e = getEvent(eventId);
        if (e == null) {
            System.out.printf("Event %s not found.%n", eventId);
            return;
        }
        int total = e.tasks.size();
        int done = 0;
        for (Task t : e.tasks) if (t.completed) done++;
        int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
        System.out.printf("\n📋 Event: %s%n", e.name);
        if (e.date != null && !e.date.isEmpty()) System.out.printf("Date: %s%n", e.date);
        if (e.description != null && !e.description.isEmpty()) System.out.printf("Description: %s%n", e.description);
        System.out.printf("Progress: %d/%d tasks done (%d%%)%n", done, total, pct);
        if (total == 0) {
            System.out.println("No tasks yet.");
            return;
        }
        System.out.println("\nTasks:");
        int i = 1;
        for (Task t : e.tasks) {
            String status = t.completed ? "[X]" : "[ ]";
            String cat = t.category != null && !t.category.isEmpty() ? " (" + t.category + ")" : "";
            System.out.printf("  %d. %s %s%s (ID: %s)%n", i++, status, t.description, cat, t.id);
        }
    }

    public void markDone(String eventId, String taskId) {
        Event e = getEvent(eventId);
        if (e == null) {
            System.out.printf("Event %s not found.%n", eventId);
            return;
        }
        for (Task t : e.tasks) {
            if (t.id.equals(taskId)) {
                if (t.completed) {
                    System.out.println("Task already completed.");
                } else {
                    t.completed = true;
                    save();
                    System.out.printf("✅ Task marked as done: %s%n", t.description);
                }
                return;
            }
        }
        System.out.printf("Task %s not found in this event.%n", taskId);
    }

    public void delete(String eventId) {
        Iterator<Event> it = events.iterator();
        while (it.hasNext()) {
            Event e = it.next();
            if (e.id.equals(eventId)) {
                String name = e.name;
                it.remove();
                save();
                System.out.printf("✅ Event deleted: %s%n", name);
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: EventOrganizer <command> [options]");
            return;
        }
        Organizer app = new Organizer();
        String cmd = args[0];
        Map<String, String> params = new HashMap<>();
        for (int i=1; i<args.length; i++) {
            if (args[i].startsWith("--") && i+1 < args.length) {
                params.put(args[i].substring(2), args[++i]);
            }
        }
        switch (cmd) {
            case "create":
                if (args.length < 2) { System.out.println("create <name> [--date DATE] [--desc TEXT]"); return; }
                app.create(args[1], params.getOrDefault("date", ""), params.getOrDefault("desc", ""));
                break;
            case "add-task":
                if (args.length < 4) { System.out.println("add-task <event_id> <description> [--category CAT]"); return; }
                app.addTask(args[1], args[2], params.getOrDefault("category", ""));
                break;
            case "list":
                app.list();
                break;
            case "show":
                if (args.length < 2) { System.out.println("show <event_id>"); return; }
                app.show(args[1]);
                break;
            case "done":
                if (args.length < 3) { System.out.println("done <event_id> <task_id>"); return; }
                app.markDone(args[1], args[2]);
                break;
            case "delete":
                if (args.length < 2) { System.out.println("delete <event_id>"); return; }
                app.delete(args[1]);
                break;
            default:
                System.out.println("Unknown command.");
        }
    }
}

File: event_organizer.go
// event_organizer.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
	"github.com/google/uuid"
)

type Task struct {
	ID          string `json:"id"`
	Description string `json:"description"`
	Category    string `json:"category"`
	Completed   bool   `json:"completed"`
}

type Event struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Date        string `json:"date"`
	Description string `json:"description"`
	Tasks       []Task `json:"tasks"`
}

type Organizer struct {
	Events []Event `json:"events"`
}

var dataFile = "events.json"

func (o *Organizer) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, o)
}

func (o *Organizer) save() {
	data, _ := json.MarshalIndent(o, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (o *Organizer) getEvent(id string) *Event {
	for i := range o.Events {
		if o.Events[i].ID == id {
			return &o.Events[i]
		}
	}
	return nil
}

func (o *Organizer) create(name, date, desc string) {
	ev := Event{
		ID:          uuid.New().String()[:8],
		Name:        name,
		Date:        date,
		Description: desc,
	}
	o.Events = append(o.Events, ev)
	o.save()
	fmt.Printf("✅ Event created: %s (ID: %s)\n", ev.Name, ev.ID)
}

func (o *Organizer) addTask(eventID, desc, category string) {
	ev := o.getEvent(eventID)
	if ev == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	task := Task{
		ID:          uuid.New().String()[:8],
		Description: desc,
		Category:    category,
	}
	ev.Tasks = append(ev.Tasks, task)
	o.save()
	fmt.Printf("✅ Task added: %s (ID: %s)\n", task.Description, task.ID)
}

func (o *Organizer) list() {
	if len(o.Events) == 0 {
		fmt.Println("No events.")
		return
	}
	fmt.Println("\n📋 Events:")
	for _, e := range o.Events {
		total := len(e.Tasks)
		done := 0
		for _, t := range e.Tasks {
			if t.Completed {
				done++
			}
		}
		pct := 0.0
		if total > 0 {
			pct = float64(done) / float64(total) * 100
		}
		fmt.Printf("  %s (ID: %s) – %d/%d tasks done (%.0f%%)\n", e.Name, e.ID, done, total, pct)
	}
}

func (o *Organizer) show(eventID string) {
	ev := o.getEvent(eventID)
	if ev == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	total := len(ev.Tasks)
	done := 0
	for _, t := range ev.Tasks {
		if t.Completed {
			done++
		}
	}
	pct := 0.0
	if total > 0 {
		pct = float64(done) / float64(total) * 100
	}
	fmt.Printf("\n📋 Event: %s\n", ev.Name)
	if ev.Date != "" {
		fmt.Printf("Date: %s\n", ev.Date)
	}
	if ev.Description != "" {
		fmt.Printf("Description: %s\n", ev.Description)
	}
	fmt.Printf("Progress: %d/%d tasks done (%.0f%%)\n", done, total, pct)
	if total == 0 {
		fmt.Println("No tasks yet.")
		return
	}
	fmt.Println("\nTasks:")
	for i, task := range ev.Tasks {
		status := "[ ]"
		if task.Completed {
			status = "[X]"
		}
		cat := ""
		if task.Category != "" {
			cat = " (" + task.Category + ")"
		}
		fmt.Printf("  %d. %s %s%s (ID: %s)\n", i+1, status, task.Description, cat, task.ID)
	}
}

func (o *Organizer) markDone(eventID, taskID string) {
	ev := o.getEvent(eventID)
	if ev == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	for i := range ev.Tasks {
		if ev.Tasks[i].ID == taskID {
			if ev.Tasks[i].Completed {
				fmt.Println("Task already completed.")
			} else {
				ev.Tasks[i].Completed = true
				o.save()
				fmt.Printf("✅ Task marked as done: %s\n", ev.Tasks[i].Description)
			}
			return
		}
	}
	fmt.Printf("Task %s not found in this event.\n", taskID)
}

func (o *Organizer) delete(eventID string) {
	idx := -1
	for i, e := range o.Events {
		if e.ID == eventID {
			idx = i
			break
		}
	}
	if idx == -1 {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	name := o.Events[idx].Name
	o.Events = append(o.Events[:idx], o.Events[idx+1:]...)
	o.save()
	fmt.Printf("✅ Event deleted: %s\n", name)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: event_organizer <command> [options]")
		return
	}
	app := &Organizer{}
	app.load()
	cmd := os.Args[1]

	switch cmd {
	case "create":
		createCmd := flag.NewFlagSet("create", flag.ExitOnError)
		name := createCmd.String("name", "", "")
		date := createCmd.String("date", "", "")
		desc := createCmd.String("desc", "", "")
		createCmd.Parse(os.Args[2:])
		if *name == "" {
			fmt.Println("create requires --name")
			return
		}
		app.create(*name, *date, *desc)

	case "add-task":
		addCmd := flag.NewFlagSet("add-task", flag.ExitOnError)
		eventID := addCmd.String("event-id", "", "")
		desc := addCmd.String("desc", "", "")
		category := addCmd.String("category", "", "")
		addCmd.Parse(os.Args[2:])
		if *eventID == "" || *desc == "" {
			fmt.Println("add-task requires --event-id and --desc")
			return
		}
		app.addTask(*eventID, *desc, *category)

	case "list":
		app.list()

	case "show":
		if len(os.Args) < 3 {
			fmt.Println("show <event_id>")
			return
		}
		app.show(os.Args[2])

	case "done":
		if len(os.Args) < 4 {
			fmt.Println("done <event_id> <task_id>")
			return
		}
		app.markDone(os.Args[2], os.Args[3])

	case "delete":
		if len(os.Args) < 3 {
			fmt.Println("delete <event_id>")
			return
		}
		app.delete(os.Args[2])

	default:
		fmt.Println("Unknown command.")
	}
}
